### PR TITLE
Compress css with nodejs

### PR DESCRIPTION
**Pull request by Morlinest**
Minify css only when building gitea so we can develop, debug and compare changes. JS should be add in the future.

Also needed for changes in building gitea with drone - including docker image with node instead of installing node on drone build.

### CODE REVIEW

File: public/css/index.css [hunk 1]
```diff
@@ -1 +1,3131 @@
...xplore{padding-top:15px;padding-bottom:80px}.explore .navbar{justify-content:center;padding-top:15px!important;margin-top:-15px!important;margin-bottom:15px!important;background-color:#fafafa!important;border-width:1px!important}.explore .navbar .octicon{width:16px;text-align:center}.ui.repository.list .item{padding-bottom:25px}.ui.repository.list .item:not(:first-child){border-top:1px solid #eee;padding-top:25px}.ui.repository.list .item .ui.header{font-size:1.5rem;padding-bottom:10px}.ui.repository.list .item .ui.header .name{word-break:break-all}.ui.repository.list .item .ui.header .metas{color:#888;font-size:14px;font-weight:400}.ui.repository.list .item .ui.header .metas span:not(:last-child){margin-right:5px}.ui.repository.list .item .time{font-size:12px;color:gray}.ui.user.list .item{padding-bottom:25px}.ui.user.list .item:not(:first-child){border-top:1px solid #eee;padding-top:25px}.ui.user.list .item .ui.avatar.image{width:40px;height:40px}.ui.user.list .item .description{margin-top:5px}.ui.user.list .item .description .octicon:not(:first-child){margin-left:5px}.ui.user.list .item .description a{color:#333}.ui.user.list .item .description a:hover{text-decoration:underline}
```
Comment (lunny) [inline public/css/index.css:1]: ~why recovery this css file?~

**Review by lafriks — CHANGES_REQUESTED** [review]
index.css file should be left minimized

File: Makefile [hunk 2]
```diff
@@ -289,20 +289,23 @@ public/js/index.js: $(JAVASCRIPTS)
...
 		exit 1; \
 	fi;
 
-.PHONY: stylesheets
-stylesheets:
+.PHONY: generate-stylesheets
+generate-stylesheets:
+	node_modules/.bin/lessc --no-ie-compat public/less/index.less public/css/index.css
+
+.PHONY: minify
```
Comment (daviian) [inline Makefile]: Maybe rename the build step to `minify-stylesheets`?

Comment (daviian) [inline Makefile]: Maybe rename this build step to `minify-stylesheets` ?

**Review by daviian — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (lafriks) [reply Makefile]: @daviian this step will later be used also to minify js files etc, so can be left as is

**Review by daviian — APPROVED** [review]

**Comment by lafriks** [comment]
I'm somehow against it. Development should be done in `.less` files and if you want to experiment you can unminify in browser

File: Makefile (hunk 2 above)
Comment (Morlinest) [reply Makefile]: @lafriks That's exact reason for that name.

**Comment by Morlinest** [comment]
@lafriks @lunny There are few reason why I think css should be unminified til build process:

1. To be able to use pre-build step with nodejs (we are using only "go" or only "nodejs" image, "generate css + minify" needs both images)
2. To be able to develop and test on local env without need to change less and regenerate css all the time
3. To be able to check existing styles and rules in development/test
4. To be able to check changes done in less file and the corresponding changes in css
5. Same workflow will be applied for JS, it is better to be able develop and check changes before minifying it

Another question: If we add JS to "minify", would you like to have 2 different workflows for JS and CSS?

**Comment by lafriks** [comment]
@Morlinest 
1. I'm not against this
2. You will still need to make changes in less files so I don't see point on making changes first in css as this is duplicate work and you anyway in the end need to regenerate css and there is chance that you can lose some of work
3. What do you mean here? (You can check rules in browser dev tools also from minimized)
4. We need to check only less changes in PR, drone ci step will verify that css file is correctly updated
5. For js there will still source files and that is only thing we need to verify, drone will check minified

**Comment by lafriks** [comment]
As for your question it should be same workflow.

File: public/css/index.css (hunk 1 above)
Comment (lunny) [reply public/css/index.css:1]: Put the unminify content on git should be friendly to develpers, so I agree with this change.

**Comment by Morlinest** [comment]
@lafriks OK, I agree with you about no need of not minified version of css.

Now look at #2581. Stylesheets test is based on differing before and after generating it from less file. But `minify` is part of go env, not js env. The only solution I see for this is to use "nodejs minify" instead.

For JS, you want to create *.min.js from *.js and use it inside template? What about development, debug?

**Comment by lunny** [comment]
LGTM

**Comment by lafriks** [comment]
Minify can be changed to nodejs version, I see no problem with that.

As for js files there should be single js file built from multiple js files. Both minified and unminified can be built and as we already have mode in app.ini (production/development) that can be used to include either minified or unminified version

**Comment by Morlinest** [comment]
I forgot about modes app.ini. I'll try do some changes.

**Comment by Morlinest** [comment]
@lafriks Changed PR to just minify css with nodejs instead in single step.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
LGTM

**Pull request merged**